pallets/registrar/src/lib.rs

```diff
@@ -419,7 +421,14 @@ pub mod pallet {
             para_id: ParaId,
             boot_nodes: BoundedVec<BoundedVec<u8, T::MaxBootNodeUrlLen>, T::MaxBootNodes>,
         ) -> DispatchResult {
-            T::RegistrarOrigin::ensure_origin(origin)?;
+            let res = T::RegistrarOrigin::ensure_origin(origin.clone());
+            if res.is_err() {
+                let origin = ensure_signed(origin)?;
+                let deposit_info = RegistrarDeposit::<T>::get(para_id).ok_or(BadOrigin)?;
```

Comment:
This is a new db read so we will need to update the benchmarks. Not sure if it's possible but we want a different weight depending on the origin. If not, the benchmark should test the worst case, which is a signed origin (instead of root).

Reply:
lets account for worst case yes

Reply:
updated the benchmark.

Reply:
I think there is a more substrate-friendly way of doing is. I propose two options:
- we use ensure_signed_or_root()  (and not registrarOrigin) for this one, and if the result is not none(), then we check that signer == depositor

- We add a new origin that is called SetBootnodesOrigin, which we can set to EitherOfDiverse<EnsureRoot, EnsureSIgned>. After this, same procedure as before would apply

Reply:
I don't think the first one is a better solution as it ignores `RegistrarOrigin`. The second solution is applicable but degrades the readability IMHO:
```rust
            let origin =
                EitherOfDiverse::<T::RegistrarOrigin, EnsureSigned<T::AccountId>>::ensure_origin(
                    origin,
                )?;

            if let sp_runtime::Either::Right(caller) = origin {
                let deposit_info = RegistrarDeposit::<T>::get(para_id).ok_or(BadOrigin)?;
                if deposit_info.creator != caller {
                    Err(BadOrigin)?;
                }
            }
```
What do you think?

Reply:
Btw, it's good to be aware of `EitherOfDiverse`, thanks!

Reply:
`RegistrarOrigin` can be changed to something like `MarkValidForCollation` origin, which basically would refer to the origin able to mark chains as valid for collation.

The problem that I see right now is that you are defining a registrarOrigin for the extrinsic, but wait, if its an error, we apply some other filter. This to me does not capture very well and its difficult to read.